ml-algorithms/src/main/java/org/opensearch/ml/engine/memory/ConversationIndexMemory.java

```diff
@@ -0,0 +1,255 @@
+    public void save(String id, Message message) {
+        this.save(id, message, ActionListener.wrap(r -> { log.info("saved message into {} memory, session id: {}", TYPE, id); }, e -> {
+            log.error("Failed to save message to memory", e);
+        }));
+    }
+
+    @Override
+    public void save(String id, Message message, ActionListener listener) {
+        mlIndicesHandler.initMemoryMessageIndex(ActionListener.wrap(created -> {
+            if (created) {
+                IndexRequest indexRequest = new IndexRequest(memoryMessageIndexName);
+                ConversationIndexMessage conversationIndexMessage = (ConversationIndexMessage) message;
+                XContentBuilder builder = XContentBuilder.builder(XContentType.JSON.xContent());
+                conversationIndexMessage.toXContent(builder, ToXContent.EMPTY_PARAMS);
+                indexRequest.source(builder);
+                client.index(indexRequest, listener);
+            } else {
+                listener.onFailure(new RuntimeException("Failed to create memory message index"));
+            }
+        }, e -> { listener.onFailure(new RuntimeException("Failed to create memory message index")); }));
+    }
+
+    public void save(Message message, String parentId, Integer traceNum, String action) {
+        this.save(message, parentId, traceNum, action, ActionListener.<CreateInteractionResponse>wrap(r -> {
+            log
+                .info(
+                    "saved message into memory {}, parent id: {}, trace number: {}, interaction id: {}",
+                    conversationId,
+                    parentId,
+                    traceNum,
+                    r.getId()
+                );
+        }, e -> { log.error("Failed to save interaction", e); }));
+    }
+
+    public void save(Message message, String parentId, Integer traceNum, String action, ActionListener listener) {
+        ConversationIndexMessage msg = (ConversationIndexMessage) message;
+        memoryManager
+            .createInteraction(
+                conversationId,
+                msg.getQuestion(),
+                null,
+                msg.getResponse(),
+                action,
+                Collections.singletonMap("None", "None"),
+                parentId,
+                traceNum,
+                listener
+            );
+    }
+
+    @Override
+    public void getMessages(String id, ActionListener listener) {
+        SearchRequest searchRequest = new SearchRequest();
+        searchRequest.indices(memoryMessageIndexName);
+        SearchSourceBuilder sourceBuilder = new SearchSourceBuilder();
+        sourceBuilder.size(10000);
+        QueryBuilder sessionIdQueryBuilder = new TermQueryBuilder(SESSION_ID, id);
+
+        BoolQueryBuilder boolQueryBuilder = new BoolQueryBuilder();
+        boolQueryBuilder.must(sessionIdQueryBuilder);
+
+        if (retrieveFinalAnswer) {
+            QueryBuilder finalAnswerQueryBuilder = new TermQueryBuilder("final_answer", true);
```

Comment:
Replace `"final_answer"` with some constant of interaction ?

Same comment for `"created_time"` of line 142

Reply:
done